### PR TITLE
add pcd map hash generator

### PR DESCRIPTION
**Pull request by taichiH**
## Description
update map_hash_generator to publish pcd map hash.

https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/142
https://github.com/tier4/autoware_api_msgs/pull/42
https://github.com/tier4/AutowareArchitectureProposal_api_msgs/pull/1
## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [pull request guidelines][pull-request-guidelines]
- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[clang-tidy-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/ClangTidyGuideline/
[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[pull-request-guidelines]: https://tier4.github.io/autoware.proj/tree/main/developer_guide/PullRequestGuideline/
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by taichiH** [comment]
Here is test result.

input path
```
$ pwd
/path/to/map_dir/pcd

$ for f in `ls | xargs`; do sha256sum $f; done
8695ee46ed912fa4d2f3ee035bfa873784cfec6792269f20f5bd0dc6e3cfda22  pointcloud_map0.pcd
8695ee46ed912fa4d2f3ee035bfa873784cfec6792269f20f5bd0dc6e3cfda22  pointcloud_map1.pcd
8695ee46ed912fa4d2f3ee035bfa873784cfec6792269f20f5bd0dc6e3cfda22  pointcloud_map2.pcd
```

result
```
ros2 launch map_launch map.launch.py pointcloud_map_path:=/path/to/map_dir/pcd map_path:=/path/to/map_dir
```
```
$ ros2 topic echo /api/autoware/get/map/info/hash --qos-reliability reliable --qos-durability transient_local
lanelet: 66fddaeb8589a91beab34e56dd90420ba61bcd2e2b2193b318ed091e2df090d1
pcd: 6b6c38b3514a563b992fb801508001243665549ebfc0a93c638478bc494eefcf
```

test
```
In [28]: text = '8695ee46ed912fa4d2f3ee035bfa873784cfec6792269f20f5bd0dc6e3cfda22'                                                 

In [29]: hashlib.sha256((text + text + text).encode()).hexdigest() == '6b6c38b3514a563b992fb801508001243665549ebfc0a93c638478bc494eefcf'                                                                                                                              
Out[29]: True
```

**Comment by isamu-takagi** [comment]
@taichiH The hashing will be executed twice If it is a file. Is that OK?
```
pcd: b05738ae46dcd374628d21c05d24da3ac99e90c722fb53b108f313afd10f8d9e
```

```
$ sha256sum pointcloud_map.pcd 
6ac84b21103b32ae43ac387d4905285442d250657f5ea100b12dfb59a1c758da  pointcloud_map.pcd
$ echo -n 6ac84b21103b32ae43ac387d4905285442d250657f5ea100b12dfb59a1c758da | sha256sum -
b05738ae46dcd374628d21c05d24da3ac99e90c722fb53b108f313afd10f8d9e  -
```

**Comment by taichiH** [comment]
@isamu-takagi 
Yes as expected. I thought the hash generation method should be the same for a single file and multiple files, so I did so.

But I just changed my mind because it's necessary to generate hash in the same way on the application side for example fms, and it feels strange.

Is it better to change the hash generation method for a single file?

**Comment by isamu-takagi** [comment]
@taichiH I think it is easier to understand.

File: map/map_loader/script/map_hash_generator [hunk 1]
```diff
@@ -54,9 +58,30 @@ class MapHashGenerator(Node):
...
+        if path.is_dir():
+            pcd_file_paths = [f for f in path.iterdir() if pathlib.Path(f).suffix == ".pcd"]
+        elif path.is_file():
+            if path.suffix == ".pcd":
+                pcd_file_paths = [path]
+        else:
+            return ""
+
+        concatenated_hash = ""
+        for pcd_file_path in pcd_file_paths:
```
Comment (isamu-takagi) [inline map/map_loader/script/map_hash_generator:76]: The paths from iterdir is in arbitrary order, so they have to be sorted.

Comment (taichiH) [reply map/map_loader/script/map_hash_generator]: @isamu-takagi fixed it in https://github.com/tier4/AutowareArchitectureProposal.iv/pull/745/commits/f08542fb191cd11df2b09b38bd300e20c3ad50ff

**Comment by taichiH** [comment]
@isamu-takagi I fixed point you've reviewed.

**Review by isamu-takagi — APPROVED** [review]
LGTM

**Pull request merged**